refactor(auth): Retire gettext string extraction, remove Mailer 'translator' dependence, new partials + section in Storybook

LZoog opened the pull request:
**You'll want to click to open several files GH doesn't open since the diff is too large**.

This commit:
* Localizes strings in `session.js` with Fluent through a new `localizeStrings` method, adds to and moves `auth.ftl` into `l10n/` and updates corresponding paths
* Renames the `location` partial to `userInfo` for more accurate naming
* Creates new `userDevice` and `userLocation` partials for use in `userInfo` rather than figuring logic and l10n in `email.js`, modifies `email.js` accordingly
* Enables us to view partials in Storybook without a layout, creates this new section with partial states
* Stops passing `translator` into Mailer instances since we no longer need it
* Creates `flattenNestedObjects` for the Renderer to localize `userDevice` and `userLocation` which are passed in as objects, removes unneeded `translatorParams` in emails
* Passes `productName` to templates instead of `productNameNew` for slight cleanup
* Removes unneeded tests, adds more tests

Because:
* We're moving away from gettext across the codebase in favor of using Fluent / our new Localizer class in at least the auth-server
* Moving `device` and `location` into partials warranted the need for a Storybook section containing these, since they display various states 

I'm going to file a few follow up tickets regardless, like:
* Remove l10n-extract, 18n-abide, Jed, translator.js, po2json, any other package/file cleanup around l10n/i18n - #12191
* I can't set `NodeRendererBindings` as the default in the constructor or constructor params for `Localizer` or `Renderer`. (Will need to check what happens again, I ran into this before but now that we're importing in multiple files it's getting annoying) - FXA-4762 / #12260

closes #11960

LZoog commented on packages/fxa-auth-server/lib/senders/email.js:
```diff
@@ -329,71 +329,18 @@ module.exports = function (log, config, bounces) {
-  Mailer.prototype._formatUserAgentInfo = function (message) {
-    // Build a first cut at a device description,
-    // without using any new strings.
-    // Future iterations can localize this better.
```
> // Future iterations can localize this better.

_7 years later_

vbudhram reviewed (commented):
@LZoog I've done a first pass on this and it looks good overall.

LZoog commented on packages/fxa-auth-server/grunttasks/ftl.js:
```diff
@@ -7,7 +7,7 @@
   grunt.config('concat', {
     ftl: {
-      src: ['lib/senders/emails/auth.ftl', 'lib/**/senders/emails/**/en.ftl'],
+      src: ['lib/l10n/auth.ftl', 'lib/**/senders/emails/**/en.ftl'],
```
I moved `auth.ftl` to live inside `lib/l10n` instead because that felt like a better spot for it since it's no longer just for emails. I also added the strings from `session.js` to live in `auth.ftl` rather than a new `en.ftl` file here.

LZoog commented on packages/fxa-auth-server/lib/routes/session.js:
```diff
@@ -506,9 +503,31 @@ module.exports = function (
+
+        // If/when we use .localizeStrings in other files, probably move where strings are
+        // maintained to separate file?
+        const titleFtlId = 'session-verify-send-push-title';
```
Cool, sounds good. We may end up using/extending `Localizer` in the future to manage our l10n for all of FxA so we might need something like that when that happens but agreed we don't really need the optimization yet.

LZoog commented on packages/fxa-auth-server/lib/senders/email.js:
```diff
@@ -1856,7 +1803,6 @@ module.exports = function (log, config, bounces) {
         productPaymentCycleNew,
         productPaymentCycleOld,
         icon: productIconURLNew,
-        product: productNameNew,
```
We don't reference `product` in any templates so I removed them here.

`translatorParams` is no longer needed since we have `flattenNestedObjects` now.

`productNameNew` was changed to `productName` for consistency - we were previously checking for (psuedo code) `if (!productName && productNameNew) { productName = productNameNew }` in our templates so we don't need that anymore now.

There's also at least one instance where I removed passing in `subscription` since we don't reference it and only need `productName`. I'm actually not sure if we ever even need to pass in the entire `subscription` object since I don't see that we ever need `planId` or `productId` in the templates.

LZoog commented on packages/fxa-auth-server/lib/senders/emails/templates/subscriptionDowngrade/en.ftl:
```diff
@@ -1,10 +1,10 @@
-# $productNameNew (String) - The name of the subscribed product, e.g. Mozilla VPN
-subscriptionDowngrade-subject = You have switched to { $productNameNew }
+# $productName (String) - The name of the subscribed product, e.g. Mozilla VPN
+subscriptionDowngrade-subject = You have switched to { $productName }
```
FWIW we are actually supposed to create new FTL IDs even if only the variable name changes, I checked 😩

However, what I can do instead is put a PR up in the l10n repo to mass replace these just before or after this PR is merged. I confirmed with cocomo that this will work and probably be best.

LZoog commented on packages/fxa-auth-server/lib/senders/renderer/index.ts:
```diff
@@ -175,4 +193,31 @@ export function splitPlainTextLine(plainText: string) {
+ * We flatten objects coming from the mailer when localizing because Fluent expects to be passed
+ * a simple object containing variable names and their values, not an object containing objects
+ *
+ * NOTE: if in the future, any template value is an _array_ of objects containing strings needing
+ * l10n, we will need to account for those variable names differently to ensure the same EJS
+ * variable matches the variable name we pass to Fluent. Right now `subscriptions` containing
+ * `productName`s is the only array of objects and since we don't localize `productName`
+ * that case isn't handled since we don't need to (yet).
```
BTW: this was the case before my changes.

LZoog commented on packages/fxa-auth-server/scripts/bulk-mailer/normalize-user-records.js:
```diff
@@ -13,37 +14,25 @@ module.exports = class UserRecordNormalizer {
-  normalizeAcceptLanguage(userRecord) {
-    // The Chinese translations were handed to us as "zh" w/o a country
-    // specified. We put these translations into "zh-cn", use "zh-cn" for
-    // Taiwan as well.
```
I ran this by Peiying as well.

I'm not sure if we just didn't support `zh-tw` for a while or if there was a bug, but it's available in the repo now and `zh-tw` users should see `zh-tw` as expected. 🤷‍♀️

LZoog commented on packages/fxa-auth-server/scripts/write-emails-to-disk.js:
```diff
@@ -121,7 +117,7 @@ function sendMail(mailer, messageToSend) {
     playStoreLink: 'https://example.com/play-store',
     invoiceNumber: '8675309',
-    cardType: 'mastercard',
+    cardType: 'MasterCard',
```
I have no idea why this ever passed. I got some failures telling me it should be `MasterCard` instead of `mastercard`.

LZoog commented on packages/fxa-auth-server/test/local/mailer_locales.js:
```diff
@@ -1,84 +0,0 @@
-    const locales = [
-      // [ locale, expected result ]
-      ['', 'en'],
-      ['db-LB', 'en'],
```
I deleted this file because we already have a bunch of similar tests in our `parseAcceptLanguage` test suite.

LZoog commented on packages/fxa-auth-server/test/local/senders/emails.ts:
```diff
@@ -2206,110 +2278,75 @@ describe('lib/senders/emails:', () => {
-      uaOSVersion: '10',
-    });
-    assert.equal(result, 'Windows 10');
+  describe('formats user agent strings sanely', () => {
```
We've also got more extensive tests for this up above when we check what the template outputs.

LZoog commented:
I just fixed up the last failures. One reason for this was because the old `parseAcceptLanguage` provided a string while the new one provides an array of strings. I updated these to `parseAcceptLanguage()[0]` to grab the first preferred locale, but we may actually want the full string like `de,it` in some spots since if a language string isn't present in the preferred locale it'll check the 2nd etc. locale before defaulting to English. I'll add a note about this in #12191

vbudhram reviewed (approved):
@LZoog Thanks!

LZoog commented on packages/fxa-auth-server/lib/senders/email.js:
```diff
@@ -147,6 +148,7 @@ module.exports = function (log, config, bounces) {
     return target;
   }
 
+  // TODO: can this be modified/removed?
```
Ah right, I'll file a follow up for this.

dschom commented on packages/fxa-auth-server/lib/l10n/index.ts:
```diff
@@ -91,26 +95,35 @@ class Localizer {
+      ftlIdMsgs.map(async (ftlIdMsg) => {
+        const { id, message } = ftlIdMsg;
+        return Promise.resolve({
+          [id]: (await l10n.formatValue(id, message)) || message,
```
Perhaps an error here should be handled? As we've been discussing somewhat recently, an unhandled error will cause promsie.all to reject immediately. It seems like one msg failing to format shouldn't prevent the others from being processed.

LZoog replied:
In this case, I don't believe it can reject. If `await l10n.formatValue(id, message)` rejects, the promise should get resolved with the original `message`. Good thinking though, I'll write a test for this to verify it's doing what I think it is.

dschom commented on packages/fxa-auth-server/lib/senders/emails/templates/subscriptionDowngrade/index.mjml:
```diff
@@ -7,15 +7,15 @@
   <mj-column>
     <mj-text css-class="text-body">
-      <span data-l10n-id="subscriptionDowngrade-content-switch" data-l10n-args="<%= JSON.stringify({productNameNew, productNameOld}) %>">You have successfully switched from <%- productNameOld %> to <%- productNameNew %>.</span>
+      <span data-l10n-id="subscriptionDowngrade-content-switch" data-l10n-args="<%= JSON.stringify({productName, productNameOld}) %>">You have successfully switched from <%- productNameOld %> to <%- productName %>.</span>
```
I noticed that Mailer.prototype.supbscriptionDownGrade provdides productNameNew, not productName old.

dschom commented on packages/fxa-auth-server/lib/senders/renderer/bindings.ts:
```diff
@@ -2,6 +2,7 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+import { string } from '@hapi/joi';
```
Unused import

LZoog replied:
Oooo good catch, oops this was from me typing "string" for Typescript and VSCode automatically imported this and I didn't realize it.

dschom reviewed (approved):
LGTM! Only nitpick is the potential added error handling inside the promise.all. There is a chance that the method would never error out in which case error handling is probably superfluous.

dschom commented on packages/fxa-auth-server/lib/l10n/index.ts:
```diff
@@ -91,26 +95,35 @@ class Localizer {
+    acceptLanguage = 'en',
+    ftlIdMsgs: FtlIdMsg[]
+  ): Promise<LocalizedStrings> {
     const { l10n } = await this.setupLocalizer(acceptLanguage);
```
I am not 100% sure, but this call might be heavier than we think. Considering this get's invoked from /session/verify/send_push, it'd be bad if this blocked the event loop in anyway. I wonder if we should double check the performance profile here. Let me know if you want to try an setup a preliminary load test.

LZoog replied:
I'm trying to think of an alternative here. `setupLocalizer` only fetches the bundles from the `acceptLanguage`, so I hope it's not too heavy. 😩 I don't think we can instantiate `Localizer` at the top of files to then immediately fetch all of the bundles for all locales (since we wouldn't have access to `acceptLanguage` yet) since we'd still have to `await` for that to finish.

I wonder 👀 We have `acceptLanguage` in [a lazy getter](https://github.com/mozilla/fxa/blob/main/packages/fxa-auth-server/lib/server.js#L201) on `app.request` so we can reference it with `request.app.acceptLanguage`. Maybe we can set up all of the Fluent bundles similarly? Would be worth checking for this kind of optimization in #12045 or even in #12191 since that may get rid of `translator` which `defineLazyGetter(request.app, 'locale'` uses.

I'm not sure in the meantime, without this kind of refactor, if there's a better way to do this. I'd be interested in what the performance profiler says either way, I keep referencing #12045 but I'd love to get a before vs. after that ticket.

pdehaan commented on packages/fxa-auth-server/lib/senders/emails/partials/userLocation/index.txt:
```diff
@@ -0,0 +1,12 @@
+<% const { city, stateCode, country } = location; %>
+  <% if (location.city) { %>
+    <% if (location.stateCode) { %>
+      location-all = "<%- location.city %>, <%- location.stateCode %> <%- location.country %> (estimated)"
```
We destructured these above on L1. Should these just be `<% if (city) { %>` (similar to L6, L9, L11 below).

LZoog replied:
I like how I did that for my own convenience and then I didn't even use it. 🥇

chenba commented:
I do have a performance concern.  It seems that there are a lot of unnecessary file IO given that, if I'm reading correctly, the message bundles are loaded afresh every time.  It's not introduced in this PR; you can see that change here: https://github.com/mozilla/fxa/commit/b0d0af958d3559b554e2b6a178b45ced9793845e#diff-400167c3d375b4bb1842b3e474a240393ac9f3907f31c8cb445e5745e085e96fL34  We could either file a follow-up or fix it in this PR.  I'll leave that up to you.  

Otherwise, LGTM.

LZoog commented:
> I do have a performance concern. It seems that there are a lot of unnecessary file IO given that, if I'm reading correctly, the message bundles are loaded afresh every time. 

@chenba Yes, this is true, and a great catch. Dan and I discussed this in another PR and we have #12045 to explore performance gains here which includes not refetching bundles every time (first bullet point). I'm going to add that ticket to a performance maintenance sprint suggestion.

LZoog commented:
Alright I pushed a couple of tweaks based on PR comments:
* Used the deconstructed values in the location partial, per @pdehaan's comment
* I wrapped the `await` in `localizeStrings` inside that `try`/`catch` as recommended by @dschom and added two more tests around these. I looked into the `formatValue` method a bit more and `Promise.reject` is never explicitly called there and it doesn't _look_  like that would reject in practice, but I'm not sure if something further down could bubble up or reject in the future so I'm fine with erring on the side of caution.

For the potential performance concerns (@chenba + @dschom), I hear you. I'm going to add #12045 to sprint 231 instead of waiting for a performance maintenance sprint. I'm not sure if we'll get to it there or the following sprint, but I'll keep an eye on it.
